In [1]:
import pandas as pd
import re

Overall Purpose
- Clean and standardize raw student data for analysis.
- Ensure numeric columns are correct and usable.
- Remove duplicates and irrelevant records.
- Prepare a clean dataset for further analysis like ranking, statistics, or visualization.

In [ ]:
df = pd.read_csv("Private_data.csv")
df.head(5)

Unnamed: 0 NAME OF THE STUDENT                   UNIVERSITY  \
0            0         Camila Wood     Stanford University, USA   
1            1  Alexander Thompson     Stanford University, USA   
2            2         Liam Taylor      Harvard University, USA   
3            3      Evelyn Jenkins     Stanford University, USA   
4            4     Michael Jackson      Harvard University, USA   
5            5         Chloe Moore     Stanford University, USA   
6            6      Nicholas Clark      Harvard University, USA   
7            7   Olivia Richardson     Stanford University, USA   
8            8        Aiden Rivera     Stanford University, USA   
9            9       Harper Turner      Harvard University, USA   
10          10         Madison Lee     Stanford University, USA   
11          11         Carter Hall      Harvard University, USA   
12          12          Mia Thomas      Harvard University, USA   
13          13         Lucas Evans                   Oxford, UK   
14          14      Abigail Bailey      Harvard University, USA   
15          15         Joseph Diaz      Harvard University, USA   
16          16      William Walker  University of Cambridge, UK   
17          17     Andrew Anderson     University of Oxford, UK   
18          18         Sophia King      Harvard University, USA   
19          19      Avery Gonzalez      Harvard University, USA   
20          20          Lily Davis      Harvard University, USA   
21          21     Scarlett Cooper     Stanford University, USA   
22          22          Grace Reed     Stanford University, USA   
23          23     Matthew Johnson      Harvard University, USA   
24          24       Jackson Baker     Stanford University, USA   

     PROGRAM NAME                                     Specialisation SEMESTER  \
0           B.Com                                            Honours      5th   
1           B.Com                                 Financial Services      5th   
2           B.Com                               Accounting Analytics      5th   
3           B.Com                                            Honours      5th   
4             MBA                             International Business      3rd   
5           B.Com                                            Honours      5th   
6             MBA                             International Business      3rd   
7             BBA                             International Business      5th   
8           B.Com                                            Honours      5th   
9             MBA                             International Business      3rd   
10            BBA                                 Business Analytics      5th   
11            MBA                             International Business      3rd   
12          B.Com   Information Technology with Accounting Analytics      5th   
13  Intg. BBA+MBA                                 Financial Services      9th   
14            MBA                             International Business      3rd   
15            MBA                             International Business      3rd   
16            MBA   Information Technology with Accounting Analytics      5th   
17  Intg. BBA+MBA                                 Financial Services      9th   
18          B.Com   Information Technology with Accounting Analytics      5th   
19            MBA  Innovation Entrepreneurship and Venture Develo...      3rd   
20            MBA  Innovation Entrepreneurship and Venture Develo...      3rd   
21            BBA                             International Business      5th   
22            BBA                             International Business      5th   
23            MBA                             International Business      3rd   
24            BBA                             International Business      5th   

     Domain  GENERAL MANAGEMENT SCORE (OUT of 50)  \
0   Finance                                    50   
1   Finance                                    5

In [ ]:
file = "Private_data"
clean_tb_name = file.lower().replace(" ", "_").replace("?", "")\
.replace(r"/", "").replace("\\", "").replace("-", "_")\
.replace(")", "").replace("(", "").replace("%", "").replace(r",", "")

Converts the file name into a safe, standardized format (all lowercase, underscores instead of spaces, removes special characters).

Prevents issues with saving/loading files on different systems or using them as variable names.

In [ ]:

clean_tb_name

'private_data'

In [ ]:
df.columns = [x.lower().replace(" ", "_").replace("?", "")\
.replace(r"/", "").replace("\\", "").replace("-", "_")\
.replace("%", "").replace(r",", "") for x in df.columns]

Standardizes column names (lowercase, underscores, no special characters).

Makes it easier to reference columns in code consistently, avoiding errors due to spaces or special characters.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 12 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   unnamed:_0                            151 non-null    int64  
 1   name_of_the_student                   151 non-null    object 
 2   university                            151 non-null    object 
 3   program_name                          151 non-null    object 
 4   specialisation                        151 non-null    object 
 5   semester                              151 non-null    object 
 6   domain                                151 non-null    object 
 7   general_management_score_(out_of_50)  151 non-null    int64  
 8   domain_specific_score_(out_50)        150 non-null    object 
 9   total_score_(out_of_100)              151 non-null    object 
 10  rank                                  151 non-null    float64
 11  percentile         

Checks the types of each column and non-null counts.

Helps identify columns that need type conversion (e.g., numeric vs string) and missing data.

In [ ]:
non_numeric_rows = df[~df['total_score_(out_of_100)'].str.replace('.', '', regex=False).str.isnumeric()]

print(non_numeric_rows)

     unnamed:_0 name_of_the_student                university program_name  \
77           77         Sophia King  Stanford University, USA          MBA   
86           86      Harper Jackson   Harvard University, USA          MBA   
139         139        Olivia White   Harvard University, USA        B.Com   

             specialisation semester domain  \
77        Digital Marketing      3rd     DM   
86   International Business      3rd     IB   
139      Business Analytics      5th     BA   

     general_management_score_(out_of_50) domain_specific_score_(out_50)  \
77                                     38                             36   
86                                     50                             22   
139                                    21                              9   

    total_score_(out_of_100)   rank  percentile  
77                         x   78.0    0.490066  
86                        -4   87.0    0.430464  
139                      aaa  140.0    0.07

Identifies rows where numeric columns (total_score_(out_of_100)) have invalid entries like x, -4, aaa.

Ensures data consistency before numeric operations; prevents errors in calculations or analysis.

In [ ]:
df['domain_specific_score_(out_50)'] = pd.to_numeric(df['domain_specific_score_(out_50)'], errors='coerce')
df['total_score_(out_of_100)'] = pd.to_numeric(df['total_score_(out_of_100)'], errors='coerce')
df['percentile'] = pd.to_numeric(df['percentile'], errors='coerce')


Forces the columns to numeric type; non-numeric entries become NaN.

Allows arithmetic and statistical operations on scores, but introduces NaN where data was invalid.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 12 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   unnamed:_0                            151 non-null    int64  
 1   name_of_the_student                   151 non-null    object 
 2   university                            151 non-null    object 
 3   program_name                          151 non-null    object 
 4   specialisation                        151 non-null    object 
 5   semester                              151 non-null    object 
 6   domain                                151 non-null    object 
 7   general_management_score_(out_of_50)  151 non-null    int64  
 8   domain_specific_score_(out_50)        149 non-null    float64
 9   total_score_(out_of_100)              149 non-null    float64
 10  rank                                  151 non-null    float64
 11  percentile         

Видаляю лічильник, так як за замовченням є нумерація

In [ ]:
df = df.drop(columns=['unnamed:_0'])

In [ ]:
duplic_rows_by_name = df.duplicated(subset='name_of_the_student')
print(df[duplic_rows_by_name])

    name_of_the_student                university   program_name  \
34         Emma Stewart  Stanford University, USA            BBA   
64          Carter Hall   Harvard University, USA  Intg. BBA+MBA   
73           Aria Evans  Stanford University, USA            BBA   
77          Sophia King  Stanford University, USA            MBA   
109         Liam Taylor   Harvard University, USA            MBA   
125          Grace Hall   Harvard University, USA            MBA   
127          Lily Davis  Stanford University, USA            BBA   
137     Scarlett Cooper  University of Oxford, UK  Intg. BBA+MBA   
138           Mia Young  Stanford University, USA            BBA   
150      Abigail Bailey   Harvard University, USA            MBA   

                                        specialisation semester   domain  \
34                                  Financial Services      5th  Finance   
64                              International Business      3rd       IB   
73                     

Ensures each student appears only once.

Prevents double-counting in analysis and statistics.

In [ ]:
df = df.drop_duplicates()

Видаляю усі рядки, що мають не той формат

In [ ]:
df = df.dropna()

Видалаяю рядки, де програма НЕ BBA або MBA, як вказано у інструкції до практичної

In [ ]:
df = df[df['program_name'].apply(lambda x: x in ['BBA', 'MBA'])]

In [ ]:
df = df.reset_index(drop=True)
df.index += 1

In [ ]:
df.head(20)

name_of_the_student                   university program_name  \
1      Michael Jackson      Harvard University, USA          MBA   
2       Nicholas Clark      Harvard University, USA          MBA   
3    Olivia Richardson     Stanford University, USA          BBA   
4        Harper Turner      Harvard University, USA          MBA   
5          Madison Lee     Stanford University, USA          BBA   
6          Carter Hall      Harvard University, USA          MBA   
7       Abigail Bailey      Harvard University, USA          MBA   
8          Joseph Diaz      Harvard University, USA          MBA   
9       William Walker  University of Cambridge, UK          MBA   
10      Avery Gonzalez      Harvard University, USA          MBA   
11          Lily Davis      Harvard University, USA          MBA   
12     Scarlett Cooper     Stanford University, USA          BBA   
13          Grace Reed     Stanford University, USA          BBA   
14     Matthew Johnson      Harvard University, USA          MBA   
15       Jackson Baker     Stanford University, USA          BBA   
16           Ava White     Stanford University, USA          BBA   
17           Mia Brown     Stanford University, USA          BBA   
18        Emma Stewart     Stanford University, USA          BBA   
19         Logan Smith     Stanford University, USA          BBA   
20     Benjamin Martin      Harvard University, USA          MBA   

                                       specialisation semester   domain  \
1                              International Business      3rd       IB   
2                              International Business      3rd       IB   
3                              International Business      5th       IB   
4                              International Business      3rd       IB   
5                                  Business Analytics      5th       BA   
6                              International Business      3rd       IB   
7                              International Business      3rd       IB   
8                              International Business      3rd       IB   
9    Information Technology with Accounting Analytics      5th       BA   
10  Innovation Entrepreneurship and Venture Develo...      3rd      E&I   
11  Innovation Entrepreneurship and Venture Develo...      3rd      E&I   
12                             International Business      5th       IB   
13                             International Business      5th       IB   
14                             International Business      3rd       IB   
15                             International Business      5th       IB   
16                             International Business      5th       IB   
17                                 Financial Services      5th  Finance   
18                                 Financial Services      5th  Finance   
19                             International Business      5th       IB   
20                             International Business      3rd       IB   

    general_management_score_(out_of_50)  domain_specific_score_(out_50)  \
1                                     50                            49.0   
2                                     49                            49.0   
3                                     50                            47.0   
4                                     49                            48.0   
5                                     47                            50.0   
6                                     50                            47.0   
7                                     47                            48.0   
8                                     49                            46.0   
9                                     48                            47.0   
10                                    49                            43.0   
11                                    49                            43.0   
12                                    49                            43.0   
13        

Перевіряємо кількість рядків

такі результати були на початку

RangeIndex: 151 entries, 0 to 150

Data columns (total 12 columns):

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 1 to 97
Data columns (total 11 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   name_of_the_student                   97 non-null     object 
 1   university                            97 non-null     object 
 2   program_name                          97 non-null     object 
 3   specialisation                        97 non-null     object 
 4   semester                              97 non-null     object 
 5   domain                                97 non-null     object 
 6   general_management_score_(out_of_50)  97 non-null     int64  
 7   domain_specific_score_(out_50)        97 non-null     float64
 8   total_score_(out_of_100)              97 non-null     float64
 9   rank                                  97 non-null     float64
 10  percentile                            97 non-null     float64
dtypes: float64(4), int64(

In [ ]:
df.to_csv('Private_data1.csv', index=False)